expression/column.go

```diff
@@ -249,8 +249,10 @@ func (col *Column) EvalString(ctx sessionctx.Context, row chunk.Row) (string, bo
 	}
 
 	val := row.GetString(col.Index)
-	if ctx.GetSessionVars().StmtCtx.PadCharToFullLength && col.GetType().Tp == mysql.TypeString {
-		valLen := len([]rune(val))
+	// See #3644.
+	// Binary type column value will changed after modify the column length.
+	if (ctx.GetSessionVars().StmtCtx.PadCharToFullLength && col.GetType().Tp == mysql.TypeString) || col.GetType().IsBinaryStr() {
```

Comment:
We store `CHAR[N]` in TiKV by removing the trailing spaces. But for `BINARY[N]`, we add trailing spaces to `N` bytes.

When modifying a column from `BINARY[X]` to `BINARY[Y]`, the underlying data should be regenerated. I think this is the correct way to fix this kind of issue.

@zimulala How do you think?

Reply:
@zz-jason 
Yes. But I think we can do this first and add a TODO. Wait until the "modify column" can handle the problem of regenerating the data and then remove the logic.

Reply:
Okay. I think we'd better file an issue about this bug.

Reply:
@zimulala Since these kinds of DDL operations are not working correctly at present. I think a workaround is to forbid these operations. Do not allow modifying a `binary(x)` column to a `binary(y)` column if `x` is not equals to `y`.

Reply:
see https://github.com/pingcap/tidb/issues/11597 for details.